### PR TITLE
Added `config-dot-notation` for config files.

### PR DESCRIPTION
**Pull request by f0zi**
In reply to https://github.com/yargs/yargs/issues/1305.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -177,6 +177,33 @@ node example.js --foo.bar
 { _: [], "foo.bar": true }
 ```
 
+_Note:_ This setting does not apply to config files (any more). See `config-dot-notation`.
```
Comment (bcoe) [inline README.md]: Perhaps:

`_Note: as of yargs-parser@14.0.0, `dot-notation` has been split into `dot-notation` and `config-dot-notation`_`

File: README.md [hunk 2]
```diff
@@ -177,6 +177,33 @@ node example.js --foo.bar
...
+
+### config-dot-notation
+
+* default: `false`
+* key: `config-dot-notation`
+
+Should keys in the config file that contain `.` split up in objects?
```
Comment (bcoe) [inline README.md]: nit:

> Should keys in the config file that contain `.` be split into objects.

File: test/yargs-parser.js [hunk 3]
```diff
@@ -567,6 +567,7 @@ describe('yargs-parser', function () {
     it('should load nested options from config file', function () {
       var jsonPath = path.resolve(__dirname, './fixtures/nested_config.json')
       var argv = parser(['--settings', jsonPath, '--nested.foo', 'bar'], {
+        configuration: { 'config-dot-notation': true },
```
Comment (bcoe) [inline test/yargs-parser.js]: I don't think this logic should need to change because we don't have keys of the form "foo.bar"?

File: test/yargs-parser.js [hunk 4]
```diff
@@ -718,6 +719,7 @@ describe('yargs-parser', function () {
 
     it('should load nested options from config object', function () {
       var argv = parser(['--nested.foo', 'bar'], {
+        configuration: { 'config-dot-notation': true },
         configObjects: [{
           a: 'a',
           nested: {
```
Comment (bcoe) [inline test/yargs-parser.js]: It would be good to add two tests, demonstrating "foo.bar" with `config-dot-notation` enabled and disabled.

File: README.md (hunk 1 above)
Comment (f0zi) [reply README.md]: Ok, I was expecting this :)

File: README.md (hunk 2 above)
Comment (f0zi) [reply README.md]: NP

File: test/yargs-parser.js (hunk 4 above)
Comment (f0zi) [reply test/yargs-parser.js]: Yeah I was thinking of adding a test but I saw "does not append nested-object keys from config to top-level key" seems to be testing exactly that.

Except, now that I look at it again, the only thing I would change is line 2100 change `'dot-notation': false` to `'config-dot-notation': false` to make it explicit.

Or would you rather like me to add another test case for this?

File: test/yargs-parser.js (hunk 3 above)
Comment (f0zi) [reply test/yargs-parser.js]: Yes, I see what you mean, but it fails without it. Ok, I'll have to look into this one more, tomorrow or Friday.

File: index.js [hunk 5]
```diff
@@ -545,7 +546,7 @@ function parse (args, opts) {
       // if the value is an inner object and we have dot-notation
       // enabled, treat inner objects in config the same as
       // heavily nested dot notations (foo.bar.apple).
-      if (typeof value === 'object' && value !== null && !Array.isArray(value) && configuration['dot-notation']) {
+      if (typeof value === 'object' && value !== null && !Array.isArray(value) && configuration['config-dot-notation']) {
```
Comment (bcoe) [inline index.js]: This logic seems wrong to me, which I believe is why you've had to explicitly turn on `config-dot-notation` to get tests passing.

What I believe we want, when `config-dot-notation` is `false` is to still run `setConfigObject` (which takes care of merging the config file with our arguments.

It's a specific edge-case you want to turn off, which is how someone handles a config file that looks like this:

```
{
  "foo.com": {
     "hello": "world"
  }
}
```

**Comment by bcoe** [comment]
just wanted to bump this PR back up, still something you'd like to see over the finish line @f0zi?

**Comment by bcoe** [comment]
@f0zi closing for now, but please feel free to come back and dust this off when you have time.
